### ISSUE TITLE
Retry jobs after ECODE_TEMP_NORES

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#996]
Originally reported of Google Code with ID 940.

```
There are many cases where a job fails with this error, asking to be retried later. Ganeti should be able to do this on its own, at least for some cases. 

In particular, when a job fails to obtain locks opportunistically, it can retry with acquiring all the locks (suppressing opportunistic locking).

A similar principle would be handy for optimistic locking, where a job acquires locks in a wrong order (like a node and then all its instances) and needs to check if nothing changed in between the locking process. If the list of acquired locks is wrong, the locking part needs to be restarted.
```
Originally added on 2014-09-10 12:39:49 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2014-09-10 12:39:52 +0000 UTC.

Now blocking: 855 (now #911);

**Comment by Ganeti-Issues-Migrator** [comment]
```
patch series just sent out for review

```
Originally added on 2014-09-23 11:59:37 +0000 UTC.

Changed State: Started

**Comment by Ganeti-Issues-Migrator** [comment]
```
commit 91b8f1207b60b308d87166b25b05c93a27b36630
Author: Klaus Aehlig <aehlig@google.com>
Date:   Fri Sep 19 11:11:27 2014 +0200

    Make mcpu handle retry requests internally
    
    If the execution of a logical unit indicates that it is
    temporarily out of resource (ECODE_TEMP_NORES), then make
    the mcpu prepare the the logical unit for retry and restart
    again.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

commit f963755896cb5edb516c7ce41981fa8f64419ff9
Author: Klaus Aehlig <aehlig@google.com>
Date:   Mon Sep 22 14:04:40 2014 +0200

    Extend base LU by PrepareRetry
    
    LUs may override this method if they know how to prepare for an internal
    retry after temporarily being out of resources.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

commit edf17b5ee9e80b6a639b45e30aa57a3f317bc631
Author: Klaus Aehlig <aehlig@google.com>
Date:   Mon Sep 22 17:04:57 2014 +0200

    Add new GenericErrror OpRetryNotSupportedError
    
    LUs that temporarily run out of resources, most notably
    instance creation with opportunistic locking, may opt
    into retrying. If they don't do so, this is the exception
    to be thrown on a retry-preparation attempt. Having a
    custom exception rather than throwing NotImplementedError
    pleases lint.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

commit 754dae725ec4f029725a76ce02ee2b9c66807ebd
Author: Klaus Aehlig <aehlig@google.com>
Date:   Mon Sep 22 15:50:39 2014 +0200

    Reset op status to wating on NotifyRetry
    
    When a job is retried, it is effectively reset, so we have to
    reset the state back to waiting again. This will also ensure
    that we're not unpleasantly surprised when NotifyStart is called
    the second time.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

commit 2a71264e675a90d2f4e2aa6372a925af6f520314
Author: Klaus Aehlig <aehlig@google.com>
Date:   Mon Sep 22 15:09:19 2014 +0200

    Extend the callback interface by NotifyRetry
    
    Given that we will support internal retries after failed opportunistic
    locking, also extend the callback interface accordingly.
    
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>
```
Originally added on 2014-09-23 14:01:13 +0000 UTC.

Changed State: Fixed

**Comment by Ganeti-Issues-Migrator** [comment]
```
Released as part of 2.14.0
```
Originally added on 2015-06-02 10:37:03 +0000 UTC.

Changed State: Released
